aws-datastore/src/main/java/com/amplifyframework/datastore/storage/sqlite/ModelTreeHelper.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.datastore.storage.sqlite;
+
+import android.database.Cursor;
+
+import com.amplifyframework.core.model.Model;
+import com.amplifyframework.core.model.ModelAssociation;
+import com.amplifyframework.core.model.ModelSchema;
+import com.amplifyframework.core.model.ModelSchemaRegistry;
+import com.amplifyframework.core.model.query.QueryOptions;
+import com.amplifyframework.core.model.query.Where;
+import com.amplifyframework.core.model.query.predicate.QueryField;
+import com.amplifyframework.core.model.query.predicate.QueryPredicate;
+import com.amplifyframework.core.model.query.predicate.QueryPredicateOperation;
+import com.amplifyframework.core.model.query.predicate.QueryPredicates;
+import com.amplifyframework.datastore.DataStoreException;
+import com.amplifyframework.datastore.storage.sqlite.adapter.SQLiteTable;
+import com.amplifyframework.util.Empty;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class to help traverse a tree of models by relationship.
+ */
+final class ModelTreeHelper {
+    private final ModelSchemaRegistry registry;
+    private final SQLiteStorageAdapter storage;
+
+    /**
+     * Constructs a model family tree traversing utility.
+     * @param registry model registry to search schema from
+     * @param storage SQLite storage engine
+     */
+    ModelTreeHelper(ModelSchemaRegistry registry,
+                    SQLiteStorageAdapter storage) {
```

Comment:
This `SQLiteStorageAdapter` looks like a bad coupling. The adapter owns a model tree helper, and the model tree helper owns a storage. But in reality, you're only ever using it to access a SQL cursor. So, a step better would be to pass just that, directly.

Reply:
Replaced storage adapter with database handle.